plugins/tiledb/io/TileDBWriter.cpp

```diff
@@ -0,0 +1,419 @@
+    args.add("compression_level", "TileDB compression level", m_compressionLevel, -1);
+}
+
+void TileDBWriter::initialize()
+{
+    if (!m_cfgFileName.empty())
+    {
+        tiledb::Config cfg(m_cfgFileName);
+        m_ctx.reset(new tiledb::Context(cfg));
+    }
+    else
+        m_ctx.reset(new tiledb::Context());
+
+    // If the array already exists on disk, throw an error
+    if (tiledb::Object::object(*m_ctx, m_arrayName).type() == tiledb::Object::Type::Array)
+        throwError("Array already exists.");
+
+    m_schema.reset(new tiledb::ArraySchema(*m_ctx, TILEDB_SPARSE));
+
+    if (!m_compressor.empty())
+    {
+        tiledb::Filter filter = createFilter(*m_ctx, m_compressor);
+        filter.set_option(TILEDB_COMPRESSION_LEVEL, m_compressionLevel);
+        m_filterList.reset(new tiledb::FilterList(*m_ctx));
+        m_filterList->add_filter(filter);
+        m_schema->set_coords_filter_list(*m_filterList);
+    }
+}
+
+void TileDBWriter::ready(pdal::BasePointTable &table)
+{
+    // get a list of all the dimensions & their types and add to schema
+    // x,y,z will be tiledb dimensions other pdal dimensions will be tiledb attributes
+    tiledb::Domain domain(*m_ctx);
+    auto layout = table.layout();
+    DimType xt = layout->findDimType("X");
+
+    if ((xt.m_type != Dimension::Type::None) &&
+            (xt.m_type == layout->findDimType("Y").m_type) &&
+                    (xt.m_type == layout->findDimType("Z").m_type))
+    {
+        auto all = layout->dims();
+        double dimMin = std::numeric_limits<double>::lowest();
+        double dimMax = std::numeric_limits<double>::max();
+
+        domain.add_dimension(tiledb::Dimension::create<double>(*m_ctx, "X", {{dimMin, dimMax}}, m_x_tile_size))
+                .add_dimension(tiledb::Dimension::create<double>(*m_ctx, "Y", {{dimMin, dimMax}}, m_y_tile_size))
+                .add_dimension(tiledb::Dimension::create<double>(*m_ctx, "Z", {{dimMin, dimMax}}, m_z_tile_size));
+
+        for (const auto& d : all)
+        {
+            std::string dimName = layout->dimName(d);
+            if ((dimName != "X") && (dimName != "Y") && (dimName != "Z"))
+            {
+                tiledb::Attribute att = createAttribute(*m_ctx, dimName, layout->dimType(d));
+
+                if (!m_compressor.empty())
+                    att.set_filter_list(*m_filterList);
+
+                m_schema->add_attribute(att);
+            }
+        }
+
+        m_schema->set_domain(domain).set_order({{TILEDB_ROW_MAJOR, TILEDB_ROW_MAJOR}});
+        m_schema->set_capacity(m_tile_capacity);
+
+        tiledb::Array::create(m_arrayName, *m_schema);
+
+        m_array.reset(new tiledb::Array(*m_ctx, m_arrayName, TILEDB_WRITE));
+        m_query.reset(new tiledb::Query(*m_ctx, *m_array));
+        m_query->set_layout(TILEDB_UNORDERED);
+    }
+    else
+    {
+        throwError("TileDB requires the X,Y,Z dimensions to be the same type.");
+    }
```

Comment:
You don't need this check.  Unless you have done something strange, X, Y and Z are always stored as doubles.  Furthermore, even if they aren't you can fetch the data as doubles, as you're doing below.

Reply:
:+1: